design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+    // a re-issuance manually for any other reason.
+    //
+    // It will be removed by the 'issuing' controller upon complete issuance.
+    CertificateConditionIssuing CertificateCondition = "Issuing"
+)
+```
+
+At the core of this proposal is the addition of the `status.iteration` field (an integer).
+This field indicates the current 'version' of the certificate.
+When a certificate is first created, the iteration is set to `1`.
+
+###
+#### Keymanager API fields
+
+We will add a new field to the `CertificateStatus` structure:
+
+```go
+package v1alpha3
+
+type CertificateStatus struct {
+    ...
+
+    // The name of the Secret resource 
+    // +optional
+    NextPrivateKeySecretName string `json:"nextPrivateKeySecretName,omitempty"`
+}
+```
+
+
+### Controller behaviour
+
+Implementing this proposal will require a complete replacement of the current
+`certificates` controller. Almost all areas of code will be replaced.
+
+The 'compare a `certificate.spec` with an x509 certificate' logic can and
+should be preserved to help maintain some semblance of backward compatibility
+for users upgrading from previous releases.
+
+#### keymanager
+
+The `keymanager` controller will be responsible for maintaining the
+`status.nextPrivateKeySecretName` field and any 'next private key' Secret
+resources that are owned by Certificates.
+
+* If the `Issuing` condition is `True`:
+    * If the `status.nextPrivateKeySecretName` field is not set:
+        * Check for existing 'next private key' Secret resources:
+            * If one matches, set the `status.nextPrivateKeySecretName` field.
+              This handles cache inconsistencies when we observe Secret
+              creation before updating `status.nextPrivateKeySecretName`.
+            * Otherwise, generate a new private key according to
+              `spec.privateKey` and store it in a new Secret resource.
+              Persist the name of the Secret as `status.nextPrivateKeySecretName`.
+    * If the `status.nextPrivateKeySecretName` field is set:
+        * If the Secret contains key data that matches `spec.privateKey`:
+            * do nothing
+        * If the Secret is not 'owned' by the Certificate, do nothing and log
+          an Event to inform the user.
+        * If the Secret is not labelled as a 'next private key', do nothing and
+          log an Event to inform the user.
+        * If the Secret does not exist:
+            * Generate a new private key according to `spec.privateKey` and
+              create a Secret resource with the given name.
+        * If the secret contains key data that does not match `spec.privateKey`:
+            * Generate a new private key according to `spec.privateKey` and
+              store it in the Secret resource.
+
+* If the `Issuing` condition is `False` or not set:
+    * Delete all owned Secret resources with the `cert-manager.io/next-private-key: "true`
+    * Ensure `status.nextPrivateKeySecretName` is unset - we may want to
```

Comment:
If the private key is meant to be rotated on each re-issuance,  I think this will need to be cleared. Otherwise the other controllers may not wait for the key to be rotated and will assume the entry is the newly minted private key? I would suggest giving users the option of specifying a key through a different mechanism and keep this field internal if what I have described is the expected behaviour of the controllers as you see it

Reply:
> If the private key is meant to be rotated on each re-issuance, I think this will need to be cleared.

I think you're right.

> I would suggest giving users the option of specifying a key through a different mechanism and keep this field internal if what I have described is the expected behaviour of the controllers as you see it

Yeah, I touched on the option of having `spec.privateKey.pinnedSecretName` below. The alternative is the `rotationPolicy` field, but we need to work out:

1) If pinnedSecretName is specified and no key is stored, should it be generated?
2) If the key data does not 'match' (i.e. wrong keySize, keyFormat etc), should it be regenerated?

I'm leaning towards using `pinnedSecretName` instead of `rotationPolicy` as I think the implied semantics of the first one are clearer.